AI-and-Analytics/End-to-end-Workloads/LidarObjectDetection-PointPillars/src/pointpillars/preprocess.cpp

```diff
@@ -0,0 +1,292 @@
+    auto min_y_range_ct13 = min_y_range_;
+    auto pillar_x_size_ct14 = pillar_x_size_;
+    auto pillar_y_size_ct15 = pillar_y_size_;
+
+    h.parallel_for(sycl::range<2>{static_cast<unsigned long>(grid_x_size_), static_cast<unsigned long>(grid_y_size_)},
+                   [=](sycl::id<2> it) {
+                     const int x = it[0];
+                     const int y = it[1];
+                     MakePillarIndexKernel(dev_pillar_count_histo_ct0, dev_counter_ct1, dev_pillar_count_ct2,
+                                           dev_x_coors, dev_y_coors, dev_x_coors_for_sub_ct5, dev_y_coors_for_sub_ct6,
+                                           dev_num_points_per_pillar, dev_sparse_pillar_map, max_num_pillars_ct9,
+                                           max_num_points_per_pillar_ct10, grid_x_size_ct11, min_x_range_ct12,
+                                           min_y_range_ct13, pillar_x_size_ct14, pillar_y_size_ct15, x, y);
+                   });
+  });
+
+  queue.memcpy(host_pillar_count, dev_pillar_count_, sizeof(int)).wait();
+
+  queue.submit([&](auto &h) {
+    auto dev_pillar_x_in_coors_ct0 = dev_pillar_x_in_coors_;
+    auto dev_pillar_y_in_coors_ct1 = dev_pillar_y_in_coors_;
+    auto dev_pillar_z_in_coors_ct2 = dev_pillar_z_in_coors_;
+    auto dev_pillar_i_in_coors_ct3 = dev_pillar_i_in_coors_;
+    auto max_num_points_per_pillar_ct11 = max_num_points_per_pillar_;
+    auto grid_x_size_ct12 = grid_x_size_;
+
+    h.parallel_for(
+        sycl::nd_range<3>(sycl::range<3>(1, 1, host_pillar_count[0]) * sycl::range<3>(1, 1, max_num_points_per_pillar_),
+                          sycl::range<3>(1, 1, max_num_points_per_pillar_)),
+        [=](sycl::nd_item<3> item_ct1) {
+          MakePillarFeatureKernel(dev_pillar_x_in_coors_ct0, dev_pillar_y_in_coors_ct1, dev_pillar_z_in_coors_ct2,
+                                  dev_pillar_i_in_coors_ct3, dev_pillar_x, dev_pillar_y, dev_pillar_z, dev_pillar_i,
+                                  dev_x_coors, dev_y_coors, dev_num_points_per_pillar, max_num_points_per_pillar_ct11,
+                                  grid_x_size_ct12, item_ct1);
+        });
+  });
+
+  queue.submit([&](auto &h) {
+    auto dev_x_coors_for_sub_ct0 = dev_x_coors_for_sub_;
+    auto dev_y_coors_for_sub_ct1 = dev_y_coors_for_sub_;
+    auto max_num_points_per_pillar_ct6 = max_num_points_per_pillar_;
+
+    h.parallel_for(
+        sycl::nd_range<3>(sycl::range<3>(1, 1, max_num_pillars_) * sycl::range<3>(1, 1, max_num_points_per_pillar_),
+                          sycl::range<3>(1, 1, max_num_points_per_pillar_)),
+        [=](sycl::nd_item<3> item_ct1) {
+          MakeExtraNetworkInputKernel(dev_x_coors_for_sub_ct0, dev_y_coors_for_sub_ct1, dev_num_points_per_pillar,
+                                      dev_x_coors_for_sub_shaped, dev_y_coors_for_sub_shaped, dev_pillar_feature_mask,
+                                      max_num_points_per_pillar_ct6, item_ct1);
+        });
+  });
```

Comment:
Where is the .wait() synchronization needed here?

Reply:
In fact, we used an in_order_queue. We updated it to an unrestricted queue now, and added wait() operations for synchronization.